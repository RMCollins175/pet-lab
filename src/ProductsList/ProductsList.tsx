interface ProductsType {
  products: ProductType[];
}

export interface ProductType {
  id: number;
  slug: string;
  title: string;
  vendor: string;
  tags: Array<string>;
  published: true;
  url: string;
  image_src: string;
  option_value: string;
  sku: string;
  price: number;
  subscription_discount: number;
  subscription: true;
}

export const ProductsList = ({ products }: ProductsType) => {
  return (
    <ul className="users-list">
      {products.map((product: ProductType) => (
        <li
          key={product.id}
          className="product-item"
          data-testID="product-list"
        >
          <span data-testID="product-name">{product.title}</span>
        </li>
      ))}
    </ul>
  );
};
