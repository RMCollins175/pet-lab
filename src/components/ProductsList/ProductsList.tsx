interface ProductsListProps {
  products: ProductType[];
}

export interface ProductType {
  id: number;
  slug: string;
  title: string;
  vendor: string;
  tags: Array<string>;
  published: true;
  url: string;
  image_src: string;
  option_value: string;
  sku: string;
  price: number;
  subscription_discount: number;
  subscription: true;
}

export const ProductsList = ({ products }: ProductsListProps) => {
  const getDiscount = (discount: any) => {
    return typeof discount === "number" ? `${discount}%` : "";
  };

  return (
    <ul className="products-list">
      <li className="product-item" data-testID="product-list">
        <div style={{ paddingLeft: "50px" }} />
        <span data-testID="product-name">
          <strong>Title</strong>
        </span>
        <span data-testID="product-price">
          <strong>Price</strong>
        </span>
        <span data-testID="subscription-discount">
          <strong>Discount</strong>
        </span>
      </li>
      {/* TODO make into Header component */}
      {products.map((product: ProductType) => (
        <li
          key={product.id}
          className="product-item"
          data-testID="product-list"
        >
          <img src={product.image_src} alt={product.title} height={50} />
          <span data-testID="product-name">{product.title}</span>
          <span data-testID="product-price">Â£{product.price}</span>
          <span data-testID="subscription-discount">
            {getDiscount(product.subscription_discount)}
          </span>
        </li>
      ))}
    </ul>
  );
};
