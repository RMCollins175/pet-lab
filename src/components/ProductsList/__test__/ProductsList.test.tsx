import { mount } from "enzyme";
import { collapseTextChangeRangesAcrossMultipleVersions } from "typescript";
import { ProductsList } from "../ProductsList";

const dummyProducts = [
  {
    id: 1,
    slug: "joint-chews-for-dogs",
    title: "Joint Care Chews",
    vendor: "PetLab",
    tags: ["Chews", "Dog"],
    published: true,
    url: "https://thepetlabco.com/products/joint-chews-for-dogs",
    image_src: "https://via.placeholder.com/300x300.png?text=Joint+Care+Chews",
    option_value: "Single Pack",
    sku: "JCSC1",
    price: 42.29,
    subscription_discount: 25,
    subscription: true,
  },
  {
    id: 2,
    slug: "skin-chews",
    title: "Dental Chews (Rawhide)",
    vendor: "PetLab",
    tags: ["Chews", "Dog"],
    published: true,
    url: "https://thepetlabco.com/products/skin-coat",
    image_src: "https://via.placeholder.com/300x300.png?text=Dental+Chews",
    option_value: "Single Pack",
    sku: "DHC1",
    price: 29.95,
    subscription_discount: 25,
    subscription: true,
  },
];

const ProductsListProps = {
  products: [],
};

const elem = <ProductsList {...(ProductsListProps as any)} />;
const wrapper = mount(elem);

describe("ProductsList component tests", () => {
  it("should render a default productList component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render some product list items based on the products data", () => {
    expect(wrapper.find('[data-testid="product-list"]').exists()).toBe(false);
    wrapper.setProps({
      products: dummyProducts,
    });
    expect(wrapper.find('[data-testid="product-list"]').exists()).toBe(true);
  });

  it("should render the correct number of products", () => {
    wrapper.setProps({
      products: dummyProducts,
    });
    expect(wrapper.find('[data-testid="product-list"]').exists()).toBe(true);
    expect(wrapper.find('[data-testid="product-list"]').length).toEqual(2);
  });

  it("should render the correct product name, price and discount values", () => {
    wrapper.setProps({
      products: dummyProducts,
    });

    expect(wrapper.find('[data-testid="product-name"]').at(0).text()).toEqual(
      "Joint Care Chews"
    );
    expect(wrapper.find('[data-testid="product-name"]').at(1).text()).toEqual(
      "Dental Chews (Rawhide)"
    );
    expect(wrapper.find('[data-testid="product-price"]').at(0).text()).toEqual(
      "£42.29"
    );
    expect(wrapper.find('[data-testid="product-price"]').at(1).text()).toEqual(
      "£29.95"
    );
    expect(
      wrapper.find('[data-testid="subscription-discount"]').at(0).text()
    ).toEqual("25%");
    expect(
      wrapper.find('[data-testid="subscription-discount"]').at(1).text()
    ).toEqual("25%");
  });
});
