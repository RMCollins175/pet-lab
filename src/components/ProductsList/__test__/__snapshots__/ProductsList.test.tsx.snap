// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList component tests should render a default productList component 1`] = `ReactWrapper {}`;
